In [14]:
import warnings
warnings.filterwarnings("ignore")

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler


In [15]:
iris = pd.read_csv("C:\\Users\\Administrator\\Desktop\\Artificial_intelligence\\datasets\\iris.csv")

In [16]:
iris.head()

petallength  petalwidth  sepallength  sepalwidth  species
0          5.1         3.5          1.4         0.2        0
1          4.9         3.0          1.4         0.2        0
2          4.7         3.2          1.3         0.2        0
3          4.6         3.1          1.5         0.2        0
4          5.0         3.6          1.4         0.2        0

In [17]:
X = iris.drop("species",axis=1)
y=iris["species"]

### 1. Standardistaion

In [18]:
X_std = StandardScaler().fit_transform(X)

In [19]:
X_std   # now th e values ranges from -1 to +1   with mean =0 and variance =1

array([[-9.00681170e-01,  1.03205722e+00, -1.34127240e+00,
        -1.31297673e+00],
       [-1.14301691e+00, -1.24957601e-01, -1.34127240e+00,
        -1.31297673e+00],
       [-1.38535265e+00,  3.37848329e-01, -1.39813811e+00,
        -1.31297673e+00],
       [-1.50652052e+00,  1.06445364e-01, -1.28440670e+00,
        -1.31297673e+00],
       [-1.02184904e+00,  1.26346019e+00, -1.34127240e+00,
        -1.31297673e+00],
       [-5.37177559e-01,  1.95766909e+00, -1.17067529e+00,
        -1.05003079e+00],
       [-1.50652052e+00,  8.00654259e-01, -1.34127240e+00,
        -1.18150376e+00],
       [-1.02184904e+00,  8.00654259e-01, -1.28440670e+00,
        -1.31297673e+00],
       [-1.74885626e+00, -3.56360566e-01, -1.34127240e+00,
        -1.31297673e+00],
       [-1.14301691e+00,  1.06445364e-01, -1.28440670e+00,
        -1.44444970e+00],
       [-5.37177559e-01,  1.49486315e+00, -1.28440670e+00,
        -1.31297673e+00],
       [-1.26418478e+00,  8.00654259e-01, -1.22754100e+00,
      

### 2. Eigen decomposition — Computing Eigenvectors and Eigenvalues

In [20]:
# for this we first need to make covariance matrix 

#### 2.1 Covariance Matrix


In [21]:
"If `rowvar` is True (default), then : each row represents a \
                                       variable, with observations in the columns. \
     Otherwise,  \
    the relationship is transposed  : each column represents a variable, while the rows  \
                                        contain observations."

cov_mat= np.cov(X_std, rowvar=False) 
cov_mat

array([[ 1.00671141, -0.11010327,  0.87760486,  0.82344326],
       [-0.11010327,  1.00671141, -0.42333835, -0.358937  ],
       [ 0.87760486, -0.42333835,  1.00671141,  0.96921855],
       [ 0.82344326, -0.358937  ,  0.96921855,  1.00671141]])

#### 2.2 Eigenvectors and Eigenvalues computation from the covariance matrix

In [23]:
cov_mat = np.cov(X_std.T) 
eig_vals, eig_vecs = np.linalg.eig(cov_mat) 
print('Eigenvectors ' ,eig_vecs) 
print('Eigenvalues ',eig_vals)

Eigenvectors  [[ 0.52237162 -0.37231836 -0.72101681  0.26199559]
 [-0.26335492 -0.92555649  0.24203288 -0.12413481]
 [ 0.58125401 -0.02109478  0.14089226 -0.80115427]
 [ 0.56561105 -0.06541577  0.6338014   0.52354627]]
Eigenvalues  [2.93035378 0.92740362 0.14834223 0.02074601]


#### 3.1 Sorting eigenvalues

In [26]:
#Make a list of (eigenvalue, eigenvector) tuples 
eig_pairs = [(np.abs(eig_vals[i]), eig_vecs[:,i]) for i in range(len(eig_vals))] 
print(type(eig_pairs)) 

<class 'list'>


In [27]:
eig_pairs

[(2.9303537755893165,
  array([ 0.52237162, -0.26335492,  0.58125401,  0.56561105])),
 (0.9274036215173417,
  array([-0.37231836, -0.92555649, -0.02109478, -0.06541577])),
 (0.1483422264816399,
  array([-0.72101681,  0.24203288,  0.14089226,  0.6338014 ])),
 (0.02074601399559571,
  array([ 0.26199559, -0.12413481, -0.80115427,  0.52354627]))]

In [30]:
#Sort the (eigenvalue, eigenvector) tuples from high to low eig_pairs.sort() 
eig_pairs.sort(reverse=True) 
print("\n",eig_pairs)


 [(2.9303537755893165, array([ 0.52237162, -0.26335492,  0.58125401,  0.56561105])), (0.9274036215173417, array([-0.37231836, -0.92555649, -0.02109478, -0.06541577])), (0.1483422264816399, array([-0.72101681,  0.24203288,  0.14089226,  0.6338014 ])), (0.02074601399559571, array([ 0.26199559, -0.12413481, -0.80115427,  0.52354627]))]


In [31]:
#Visually confirm that the list is correctly sorted by decreasing eigenvalues 
print('\n\n\nEigenvalues in descending order:') 
for i in eig_pairs: 
    print(i[0])




Eigenvalues in descending order:
2.9303537755893165
0.9274036215173417
0.1483422264816399
0.02074601399559571


#### 3.2 Explained Variance

In [35]:
tot = sum(eig_vals) 

print("\n",tot) 

var_exp = [(i / tot)*100 for i in sorted(eig_vals, reverse=True)] 
print("\n1. Variance Explained\n",var_exp) 

cum_var_exp = np.cumsum(var_exp) 
print("\n\n2. Cumulative Variance Explained\n",cum_var_exp) 

print("\n\n3. Percentage of variance the first two principal components each contain\n ", \
      var_exp[0:2]) 

print("\n\n4. Percentage of variance the first two principal components together contain\n", \
      sum(var_exp[0:2]))


 4.026845637583894

1. Variance Explained
 [72.77045209380134, 23.030523267680643, 3.683831957627389, 0.5151926808906266]


2. Cumulative Variance Explained
 [ 72.77045209  95.80097536  99.48480732 100.        ]


3. Percentage of variance the first two principal components each contain
  [72.77045209380134, 23.030523267680643]


4. Percentage of variance the first two principal components together contain
 95.80097536148199


#### 4. Construct the projection matrix W from the selected k eigenvectors

In [36]:
print(eig_pairs[0][1]) 
print(eig_pairs[1][1]) 
matrix_w = np.hstack((eig_pairs[0][1].reshape(4,1), eig_pairs[1][1].reshape(4,1))) 
#hstack: Stacks arrays in sequence horizontally (column wise)

[ 0.52237162 -0.26335492  0.58125401  0.56561105]
[-0.37231836 -0.92555649 -0.02109478 -0.06541577]


In [37]:
matrix_w

array([[ 0.52237162, -0.37231836],
       [-0.26335492, -0.92555649],
       [ 0.58125401, -0.02109478],
       [ 0.56561105, -0.06541577]])

### Projection on to the dataset

In [43]:
Y = X_std.dot(matrix_w) 
principalDf = pd.DataFrame(data = Y , columns = ['principal component 1', 'principal component 2']) 
principalDf.head()

principal component 1  principal component 2
0              -2.264542              -0.505704
1              -2.086426               0.655405
2              -2.367950               0.318477
3              -2.304197               0.575368
4              -2.388777              -0.674767

### Final Data Frame

In [41]:
finalDf = pd.concat([principalDf,pd.DataFrame(y,columns = ['species'])], axis = 1) 
finalDf.head()

principal component 1  principal component 2  species
0              -2.264542              -0.505704        0
1              -2.086426               0.655405        0
2              -2.367950               0.318477        0
3              -2.304197               0.575368        0
4              -2.388777              -0.674767        0

### Conclusion:
    1. first two principal components contain 95.80% of the information. 
    
    2. The first principal component contains 72.77% of the variance and the second principal                                                component contains 23.03% of the variance.

### We can directly use the Pca library of the Python

In [44]:
from sklearn.decomposition import PCA

pca = PCA(0.95) # n_components= no.of components required.
principalComponents = pca.fit_transform(X_std) 
principalDf1 = pd.DataFrame(data = principalComponents , columns = ['principal component 1', 'principal component 2'])
principalDf1.head(5) 

principal component 1  principal component 2
0              -2.264542               0.505704
1              -2.086426              -0.655405
2              -2.367950              -0.318477
3              -2.304197              -0.575368
4              -2.388777               0.674767